packages/clay-core/src/tree-view/useTree.ts

```diff
@@ -0,0 +1,183 @@
+}
+
+export interface ITreeState<T> extends Pick<ICollectionProps<T>, 'items'> {
+	expandedKeys: Set<Key>;
+	open: (key: Key) => void;
+	reorder: (from: Array<number>, path: Array<number>) => void;
+	selection: IMultipleSelectionState;
+	toggle: (key: Key) => void;
+}
+
+export function useTree<T>(props: ITreeProps<T>): ITreeState<T> {
+	const [expandedKeys, setExpandedKeys] = useInternalState<Set<Key>>({
+		initialValue: new Set(),
+		onChange: props.onExpandedChange,
+		value: props.expandedKeys,
+	});
+
+	const [items, setItems] = useInternalState({
+		initialValue: [],
+		onChange: props.onItemsChange,
+		value: props.items,
+	});
+
+	const selection = useMultipleSelection({
+		onSelectionChange: props.onSelectionChange,
+		selectedKeys: props.selectedKeys,
+	});
+
+	const reorder = (from: Array<number>, path: Array<number>) => {
+		if (!Array.isArray(items)) {
+			return;
+		}
+
+		const tree = createImmutableTree(items);
+
+		tree.produce(from, path);
+
+		setItems(tree.applyPatches());
+	};
+
+	const toggle = (key: Key) => {
+		const expanded = new Set(expandedKeys);
+
+		if (expanded.has(key)) {
+			expanded.delete(key);
+		} else {
+			expanded.add(key);
+		}
+
+		setExpandedKeys(expanded);
+	};
+
+	const open = (key: Key) => {
+		const expanded = new Set(expandedKeys);
+
+		if (!expanded.has(key)) {
+			expanded.add(key);
+		}
+
+		setExpandedKeys(expanded);
+	};
+
+	return {
+		expandedKeys,
+		items,
+		open,
+		reorder,
+		selection,
+		toggle,
+	};
+}
+
+// RFC 6902 4.4
+type Patch = {
+	op: 'move';
+	from: Array<number>;
+	path: Array<number>;
+};
+
+function createImmutableTree<T extends Array<Record<string, any>>>(tree: T) {
+	const patches: Array<Patch> = [];
+
+	function nodeByPath(path: Array<number>): any {
+		const queue = [...path];
+		const rootIndex: number = queue.shift() as number;
+		let item = tree[rootIndex];
+		let parent = tree;
+		let index = rootIndex;
+
+		while (queue.length) {
+			index = queue.shift() as number;
+
+			if (item.children) {
+				parent = item.children;
+				item = item.children[index];
+			}
+		}
+
+		return {
+			index,
+			item,
+			parent,
+		};
+	}
+
+	function applyPatches(): T {
+		patches.forEach((patch) => {
+			const {from, op, path} = patch;
+
+			switch (op) {
+				// Applies the operation on the tree, the move is functionally
+				// identical to a "remove" operation on the `from` location and
+				// immediately followed by the "add" operation at the target
+				// location with the value that was removed.
+				case 'move': {
+					const nodeToRemove = nodeByPath(from);
+					const parentNode = nodeToRemove.parent.filter(
+						(item: any, index: number) =>
+							index !== nodeToRemove.index
+					);
+
+					tree[from[0]] = {
+						...tree[from[0]],
+						children: parentNode,
+					};
+
+					const {item} = nodeByPath(path);
+
+					tree[path[0]] = {
+						...item,
+						children: [...item.children, nodeToRemove.item],
+					};
```

Comment:
It was very simple, good work here too. I will try to further explore the idea of making a copy of the structure so as not to manipulate the original structure. I'll explore some ideas like not needing to create a copy of each element but only the path that is being modified, kind of a partial copy. I'll try something like this.

Reply:
Yes, very simple and that's what I liked about it. I'll also have a thought about what you're just said (partial copy) sounds intersting.

Reply:
Yeah, I was researching more about this, instead of making a complete copy of the structure, we would keep the root part of the tree a copy and the nested items that correspond to the path would be a copy too but the rest would still be a reference to the objects of the original tree, so in a way we're not modifying the original structure.

An interesting thing with this is that maybe we can use the same flow to get the paths (`nodeByPath` here) to do that, so maybe we don't need to make another iterable. This is more in theory but I can be totally wrong 😅.

I think if we get that, it's going to be awesome, because we deal with it immutable, we don't use a lot of memory to make a copy every time we modify the tree. I'm going to explore that today, I'll bring you more information whatever.

Reply:
> That's an interesting idea, but when you mentioned the word "reference", I still see it as "mutating" something

Yeah, in a way. Because we still have a reference to the path but we haven't modified it we keep it intact, it means we have the base of the tree on the first level being a copy and the nested ones are references but the path to be modified will be a copy, so even if you're making a change this is not reflected in the original tree, this would be a semi persistence of the structure perhaps.

We are still dealing with immutability the original structure is immutable but the new structure is created with the modifications has memory-level references to the original object paths and only a part of this new structure that is new.

![image](https://user-images.githubusercontent.com/13750819/132736490-43863c8d-e183-4eeb-8624-d158f12cf9d4.png)

This is similar to this image that can give a view about it, the blue corresponds to the path being modified, ie in this case adding the `e` and creating the new structure based on the original with the name `ys`. In the path in the `path` are copies and the rest are just references to the original tree.

Reply:
Well, I spent the afternoon looking at it but I didn't make much progress, I found some documents and articles on semi-persistent structures but nothing to help with that yet. I think I'll have to explore more coding to test how it would be done and refine the code. I'm also thinking that to do this I would need some auxiliary structure to help mount the tree.

Reply:
> I'm also thinking that to do this I would need some auxiliary structure to help mount the tree.

Yeah that makes sense to me

Concerning the references, I understood what you meant but I was thinking of "references" (and pointers), like in C, i.e. 

```c
int x = 5;
int *px = &x;
*px = 666;
```

anyway let's forget my original comment

Reply:
Yes, perfect, they are pointers.